table/block_based_table_reader.cc

```diff
@@ -1785,32 +1792,53 @@ BlockBasedTable::PartitionedIndexIteratorState::NewSecondaryIterator(
 //
 // REQUIRES: this method shouldn't be called while the DB lock is held.
 bool BlockBasedTable::PrefixMayMatch(const Slice& internal_key,
-                                     const SliceTransform* prefix_extractor) {
+                                     const ReadOptions& read_options,
+                                     const SliceTransform* prefix_extractor,
+                                     const bool prefix_extractor_changed) {
   if (!rep_->filter_policy) {
     return true;
   }
 
-  assert(prefix_extractor != nullptr);
+  const SliceTransform* prefix_extractor_to_use;
+
+  if (rep_->table_prefix_extractor == nullptr) {
+    if (prefix_extractor_changed) {
+      return true;
+    }
+    prefix_extractor_to_use = prefix_extractor;
+  }
+  else {
+    prefix_extractor_to_use = rep_->table_prefix_extractor.get();
+  }
   auto user_key = ExtractUserKey(internal_key);
-  if (!prefix_extractor->InDomain(user_key)) {
+  if (!prefix_extractor_to_use->InDomain(user_key)) {
     return true;
   }
-  assert(rep_->table_properties->prefix_extractor_name.compare(
-             prefix_extractor->Name()) == 0);
-  auto prefix = prefix_extractor->Transform(user_key);
 
   bool may_match = true;
   Status s;
 
   // First, try check with full filter
-  auto filter_entry = GetFilter(prefix_extractor);
+  auto filter_entry = GetFilter(prefix_extractor_to_use);
   FilterBlockReader* filter = filter_entry.value;
+  bool filter_checked = true;
   if (filter != nullptr) {
+    // set prefix using prefix_extractor from the SST
+    auto prefix = prefix_extractor_to_use->Transform(user_key);
     if (!filter->IsBlockBased()) {
       const Slice* const const_ikey_ptr = &internal_key;
-      may_match = filter->PrefixMayMatch(prefix, prefix_extractor, kNotValid,
-                                         false, const_ikey_ptr);
+      may_match = filter->RangeMayExist(
+          read_options, user_key, prefix_extractor_to_use,
+          rep_->internal_comparator.user_comparator(), prefix, kNotValid,
+          false, const_ikey_ptr, &filter_checked, prefix_extractor_changed);
```

Comment:
Try to eliminate number of arguments of RangeMayExist(). Options like kNotValid, false should not be passed.